Add a ControlNet model & pipeline (#2407)
huggingface/diffusers: docs/source/en/api/pipelines/stable_diffusion/control_net.mdx

<!--Copyright 2023 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# Text-to-Image Generation with ControlNet guidance

## StableDiffusionControlNetPipeline

ControlNet by [@lllyasviel](https://huggingface.co/lllyasviel) is a neural network structure to control diffusion models by adding extra conditions.

It has integration with Stable Diffusion and 8 pre-trained models that conditions the models on different attributes 
(such as edge detection, scribbles, depth maps, semantic segmentations and more)

The original codebase can be found here: 
- [lllyasviel/ControlNet](https://github.com/lllyasviel/ControlNet)

Available Checkpoints are:

> - TODO: fill here.

Author: Add checkpoint list at https://huggingface.co/takuma104 ?

<!--Copyright 2023 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

> # Text-to-Image Generation with ControlNet guidance

Reviewer: ```suggestion
# Text-to-Image Generation with ControlNet Conditioning
```

## StableDiffusionControlNetPipeline

ControlNet by [@lllyasviel](https://huggingface.co/lllyasviel) is a neural network structure to control diffusion models by adding extra conditions.

> It has integration with Stable Diffusion and 8 pre-trained models that conditions the models on different attributes 

Reviewer: ```suggestion
There are 8 pre-trained ControlNet models that were trained to condition the original Stable Diffusion model on different inputs, 
```

> (such as edge detection, scribbles, depth maps, semantic segmentations and more)

Reviewer: ```suggestion
such as edge detection, scribbles, depth maps, semantic segmentations and more.

Using the pretrained models we can provide control images (for example, a depth map) to control Stable Diffusion text-to-image generation so that it follows the structure of the depth image and fills in the details.
```

The original codebase/paper can be found here: 
- [Code](https://github.com/lllyasviel/ControlNet)
- [Paper](https://arxiv.org/abs/2302.05543)

## Available checkpoints

ControlNet requires a *control image* in addition to the text-to-image *prompt*. 

> Each pretrained model is trained using a different method. 

Reviewer: ```suggestion
Each pretrained model is trained using a different conditioning method that requires different conditioning images. For example, Canny edge conditioning requires the control image to be the output of a Canny filter, while depth conditioning requires the control image to be a depth mask.
```

> For this reason, the image content passed to the `image` argument of `pipe()` is different. 

Reviewer: ```suggestion
```

See the overview and image examples.

All checkpoints are converted from [lllyasviel/ControlNet](https://huggingface.co/lllyasviel/ControlNet).

> ### ControlNet + SD1.5

Reviewer: ```suggestion
### ControlNet + Stable Diffusion 1.5
```

Reviewer: ```suggestion
### ControlNet + Stable Diffusion 1.5

These ControlNet models were trained to condition the [Stable Diffusion v1.5 model](https://huggingface.co/runwayml/stable-diffusion-v1-5) published by RunwayML.
```

|[takuma104/control_sd15_openpose](https://huggingface.co/takuma104/control_sd15_openpose)<br/> *Trained with OpenPose bone image*  |A [OpenPose bone](https://github.com/CMU-Perceptual-Computing-Lab/openpose) image.|<a href="https://huggingface.co/takuma104/controlnet_dev/blob/main/gen_compare/control_images/converted/control_human_openpose.png"><img width="64" src="https://huggingface.co/takuma104/controlnet_dev/resolve/main/gen_compare/control_images/converted/control_human_openpose.png"/></a>|
|[takuma104/control_sd15_scribble](https://huggingface.co/takuma104/control_sd15_scribble)<br/> *Trained with human scribbles*  |A hand-drawn monochrome image with white outlines on a black background.|<a href="https://huggingface.co/takuma104/controlnet_dev/blob/main/gen_compare/control_images/converted/control_vermeer_scribble.png"><img width="64" src="https://huggingface.co/takuma104/controlnet_dev/resolve/main/gen_compare/control_images/converted/control_vermeer_scribble.png"/></a>|

> |[takuma104/control_sd15_seg](https://huggingface.co/takuma104/control_sd15_seg)<br/>*Trained with semantic segmentation*  |An [ADE20K](https://groups.csail.mit.edu/vision/datasets/ADE20K/)'s segmentation protocol image.|<a href="https://huggingface.co/takuma104/controlnet_dev/blob/main/gen_compare/control_images/converted/control_room_seg.png"><img width="64" src="https://huggingface.co/takuma104/controlnet_dev/resolve/main/gen_compare/control_images/converted/control_room_seg.png"/></a>|

Reviewer: This table is great! Would it make sense to maybe add another thumbnail with an example generation?

Author: Thanks! Indeed. I added generated examples. https://github.com/huggingface/diffusers/pull/2407/commits/161aac28dc6b1c638e9cd95396ed6e6cbbf4eea4

## Resources

- [Colab Notebook Example](https://colab.research.google.com/drive/1AiR7Q-sBqO88NCyswpfiuwXZc7DfMyKA?usp=sharing)
- [controlnet_hinter](https://github.com/takuma104/controlnet_hinter): Image Preprocess Library for ControlNet

## Usage example

> - Basic Example (Canny Edge)

Reviewer: ```suggestion
- Basic Example (Canny Edge)

The conditioning image is an outline of the image edges, as detected by a Canny filter. This is the example we'll use to control the generation:

![White on black edges detected on Vermeer's Girl with a Pearl Earring portrait](https://huggingface.co/takuma104/controlnet_dev/resolve/main/vermeer_canny_edged.png)

```

```python
from diffusers import StableDiffusionControlNetPipeline
from diffusers.utils import load_image

# Canny edged image for control
canny_edged_image = load_image("https://huggingface.co/takuma104/controlnet_dev/resolve/main/vermeer_canny_edged.png")
pipe = StableDiffusionControlNetPipeline.from_pretrained("takuma104/control_sd15_canny").to("cuda")
image = pipe(prompt="best quality, extremely detailed", image=canny_edged_image).images[0]
image.save("generated.png")
```

> 

Reviewer: ```suggestion

Note that the text prompt does not make any reference to the structure or contents of the image we are generating. Stable Diffusion interprets the control image as an additional input that controls what to generate.
```

Also, can we show the result here?

> - Using SD1.x variant model to control

Reviewer: ```suggestion
- Controlling custom Stable Diffusion 1.5 models

In the following example we use PromptHero's [Openjourney model](https://huggingface.co/prompthero/openjourney), which was fine-tuned from the base Stable Diffusion v1.5 model on images from Midjourney. This model has the same structure as Stable Diffusion 1.5 but is capable of producing a different output style.
```

```py
from diffusers import StableDiffusionControlNetPipeline, AutoencoderKL, UNet2DConditionModel
from diffusers.utils import load_image

# Canny edged image for control
canny_edged_image = load_image("https://huggingface.co/takuma104/controlnet_dev/resolve/main/vermeer_canny_edged.png")

base_model_id = "prompthero/openjourney"  # an example: openjourney model

> vae = AutoencoderKL.from_pretrained(base_model_id, subfolder="vae").to("cuda")

Reviewer: Question: is the vae from this model different to the one in Stable Diffusion 1.5? If it is, how are the conditioning latents compatible with it?

Author: The ControlNet VAE is probably the same as the one in SD1.5 (I will try to compare the binaries later).

Although it wasn't the most accurate experiment, when I tried to override only UNet and when I tried to override both UNet and VAE, the only difference I saw was with [Anything](https://huggingface.co/takuma104/controlnet_dev/blob/main/model_compare/anythin-v3.0.md). ([OpenJourney](https://huggingface.co/takuma104/controlnet_dev/blob/main/model_compare/openjourney.md), [Nitro-Diffusion](https://huggingface.co/takuma104/controlnet_dev/blob/main/model_compare/Nitro-Diffusion.md) results) 

I saw that Anything has a custom VAE. There are such models, so I think this sample code is fine as is.

unet = UNet2DConditionModel.from_pretrained(base_model_id, subfolder="unet").to("cuda")

pipe = StableDiffusionControlNetPipeline.from_pretrained("takuma104/control_sd15_canny", unet=unet, vae=vae).to("cuda")
image = pipe(prompt="best quality, extremely detailed", image=canny_edged_image, width=512, height=512).images[0]
image.save("generated.png")
```

> 

Reviewer: Let's show the result here if we can, so the reader can see the difference.

Author: I agree! I tried using the same seed and settings with OpenJourney, but the differences were not as clear as I had expected. I also tried with other models, but only Anything, the anime model, showed clear differences. I thought Anything would become more clearly anime-like, but it wasn't quite so. Maybe it's a characteristic of the Canny model, so I'm thinking of generating images with other ControlNet models to see if there are differences.

https://huggingface.co/takuma104/controlnet_dev/blob/main/model_compare/README.md

Reviewer: The `Anything` model is clearly different, yes. Another option could be https://huggingface.co/22h/vintedois-diffusion-v0-1 (haven't tested it myself).

Author: Looks like a nice model! I will try this one as well.

Author: I have added `vintedois-diffusion` as a new Unet model. It may not have as distinctive features as Anything, but it still stands out.
https://huggingface.co/takuma104/controlnet_dev/blob/main/model_compare/README.md

I tried other ControlNet models to see if the Canny model had strong constraints on the image. Looking at these results, Depth and OpenPose seem to be good options. If you want to use Vermeer's painting, OpenPose's pose estimation may not be very effective, so Depth may be the best option. What do you think?
https://huggingface.co/takuma104/controlnet_dev/blob/main/model_compare_full/README.md

`Normal` looks a little strange because the RGB conversion failed.
